### PR TITLE
Hawq-473. Implement adding a entry into gp_configuration_history and  add a description column to show the reason if this segment is down

### PR DESCRIPTION
**Pull request by linwen**
Please review. 
Thanks!

### CODE REVIEW

File: src/backend/resourcemanager/resourcepool.c [hunk 1]
```diff
@@ -844,22 +936,38 @@ int addHAWQSegWithSegStat(SegStat segstat, bool *capstatchanged)
...
 
 		/*
-		 * This is a new host registration. Normally the status is available,
-		 * But if the number of failed temporary directory exceeds guc,
+		 * If in Yarn mode, the new registered segment is marked
+		 * as DOWN, since master hasn't get a cluster report for it from YARN.
+		 */
+		if (DRMGlobalInstance->ImpType != NONE_HAWQ2)
+		{
+			segresource->Stat->StatusDesc |= SEG_STATUS_NO_YARN_NODE_REPORT;
```
Comment (jiny2) [inline src/backend/resourcemanager/resourcepool.c]: suggest to change to SET_STATUS_NO_GRM_NODE_REPORT , in the future, maybe more global resource managers are supported.

File: src/backend/resourcemanager/resourcepool.c [hunk 2]
```diff
@@ -901,37 +1019,58 @@ int addHAWQSegWithSegStat(SegStat segstat, bool *capstatchanged)
...
+			 * and RM reset flag in StatusDesc
+			 */
+			if ((segresource->Stat->StatusDesc & SEG_STATUS_HEARTBEAT_TIMEOUT) != 0)
+				segresource->Stat->StatusDesc &= ~SEG_STATUS_HEARTBEAT_TIMEOUT;
+			if ((segresource->Stat->StatusDesc & SEG_STATUS_RUALIVE_FAILED) != 0)
+				segresource->Stat->StatusDesc &= ~SEG_STATUS_RUALIVE_FAILED;
+			if ((segresource->Stat->StatusDesc & SEG_STATUS_COMMUNICATION_ERROR) != 0)
+				segresource->Stat->StatusDesc &= ~SEG_STATUS_COMMUNICATION_ERROR;
+			if ((segresource->Stat->StatusDesc & SEG_STATUS_RM_RESET) != 0)
+				segresource->Stat->StatusDesc &= ~SEG_STATUS_RM_RESET;
```
Comment (jiny2) [inline src/backend/resourcemanager/resourcepool.c]: suggest add debug log for status changes

File: src/backend/resourcemanager/resourcepool.c [hunk 3]
```diff
@@ -4678,3 +4826,73 @@ void dumpResourcePoolHosts(const char *filename)
 
     fclose(fp);
 }
+
+#define	SEG_STATUS_HEARTBEAT_TIMEOUT			0x00000001
```
Comment (jiny2) [inline src/backend/resourcemanager/resourcepool.c]: Move to header file resourcepool.h

**Comment by jiny2** [comment]
+1 The code looks good to me.

**Comment by huor** [comment]
+1

**Pull request closed without merging**